### ISSUE TITLE
I wanted to hide the keyboard manually, without altering the library, but it is not possible because the dismiss by tap is triggered in the `ModalBottomSheet` and can't be hooked into.

### ISSUE THREAD

**Issue by alexstyl** [composablehorizons/compose-unstyled#110]
I wanted to hide the keyboard manually, without altering the library, but it is not possible because the dismiss by tap is triggered in the `ModalBottomSheet` and can't be hooked into.

If there was a callback executed before dismiss, then the library user could hide the keyboard there.

> Do you know how it works? is the LocalSoftwareKeyboardController per window?

From the experiments, it seems so, I too don't quite understand this.

_Originally posted by @Radiokot in https://github.com/composablehorizons/compose-unstyled/issues/109#issuecomment-3196268558_

**Comment by alexstyl** [comment]
Found the culprint. https://x.com/alexstyl/status/1957643759442882767

The keyboard animation seems to be linked to the `android:windowAnimationStyle` of the dialog theme.

Used what seems to be the platform default (`@android:style/Animation.Dialog`) but it looks off.

Still investigating what the value should be so it looks right.

https://github.com/user-attachments/assets/26877835-c607-4032-bb5d-c31d03463f93

**Comment by alexstyl** [comment]
This is now fixed and be available in the next release.

This is fixed in 2 parts.

1) On Android 36, it seems that the IME animation is linked to the window exit animation. Setting that to an 'empty' animation fixes the issue. However, it seems that the ime dissapears instantly on Android 30 to Android 35. Nothing we can do in order to automatically animate it away. Luckily we can hide the keyboard programmatically thanks to (2)

2) Call `onDismiss` also when the user touches outside the sheet, or when pressing Esc/back. This way you have a callback to hide the keyboard when you need.

Btw, it seems that the `LocalSoftwareKeyboardController` is indeed tied to the window. So in order to hide the soft keyboard from the callback you will need to do:

```kotlin
        var currentKeyboardController = LocalSoftwareKeyboardController.current

        ModalBottomSheet(state = modalSheetState, onDismiss = { currentKeyboardController?.hide() }) {
            currentKeyboardController = LocalSoftwareKeyboardController.current
        }
```

https://github.com/user-attachments/assets/3ac54f5c-4d86-45fa-8f99-1ac71dbf58c6

**Comment by Radiokot** [comment]
Thank you very much, Alex. You've done a deep investigation on this issue! The animation on the latest video looks delightful.

**Comment by alexstyl** [comment]
For sure. I'd rather have stuff like this included in the lib so you can focus on your app, instead of every developer out there having to deal with such issues.

**Comment by alexstyl** [comment]
@Radiokot just to be safe, could you get `main` and see if it fixes your issue locally?

**Comment by Radiokot** [comment]
Yes, I'll do it now.

**Comment by Radiokot** [comment]
It works and looks the best with a `tween(200)` sheet animation spec.

By the way, in order to build the demo from `main`, I had to locally add
```
implementation("com.google.android.material:material:1.12.0")
```
to the `androidMain` dependencies, otherwise it doesn't build due to unresolved style references in the XML.

**Comment by alexstyl** [comment]
which demo (there are two)? I don't get any issues here

**Comment by Radiokot** [comment]
It was an issue at `bcc41ad`:
```
ERROR: AAPT: error: resource style/Theme.Material3.DayNight (aka com.composeunstyled.demo:style/Theme.Material3.DayNight) not found.
error: resource style/TextAppearance.Material3.HeadlineLarge (aka com.composeunstyled.demo:style/TextAppearance.Material3.HeadlineLarge) not found.
error: failed linking references.
```
But it is no more. Never mind.

**Comment by alexstyl** [comment]
All good. there was an issue before and sorted it out. thanks for checking
